docusaurus/docs/e2e-test-a-plugin/test-a-data-source-plugin/custom-variable-editor.md

```diff
@@ -0,0 +1,95 @@
+keywords:
+  - grafana
+  - plugins
+  - plugin
+  - testing
+  - e2e
+  - data-source
+  - variable editor
+sidebar_position: 30
+---
+
+## Introduction
+
+[Variable queries](https://grafana.com/docs/grafana/latest/dashboards/variables/add-template-variables/#add-a-query-variable) allow users to query the data source to load lists of for example metric names. The variable can then be referenced in queries, making dashboards more interactive and dynamic. If your data source plugin implements the custom variable support API, you may want to use the `variableEditPage` fixture to test that your plugin's variable implementation works as expected.
+
+### Testing the custom variable editor
+
+In the following example, we're testing that the custom variable editor renders a certain field when the `ListByDimensions` query type is chosen.
+
+```ts title="customVariableEditor.spec.ts"
+test('should display Dimensions field only if ListByDimensions is selected', async ({
+  variableEditPage,
+  page,
+  readProvisionedDataSource,
+}) => {
+  const ds = await readProvisionedDataSource({ fileName: 'datasources.yaml' });
+  await variableEditPage.setVariableType('Query');
+  await variableEditPage.datasource.set(ds.name);
+  const dimensionField = variableEditPage.getByTestIdOrAriaLabel('Dimensions');
+  await expect(dimensionField).not.toBeVisible();
+  await variableEditPage.getByLabel('Query type').fill('ListByDimensions');
+  await page.keyboard.press('Enter');
+  await expect(dimensionField).toBeVisible();
+});
+```
+
+### Testing the variable query execution flow
+
+In the next example, we're doing an integration test where the entire variable query data flow of the plugin is tested. For successful variable queries, the resulting options are displayed at the bottom of the variable edit page. You can use the matcher `toDisplayPreviews` to assert that the expected previews are being displayed.
+
+![](/img/variable-preview.png)
+
+:::warning
+Hitting third-party APIs in end-to-end tests can be useful, but it may also have security concerns as well as introduce flakiness to your CI pipeline. You should always think it through carefully and consider using mocks instead.
+:::
+
+```ts title="customVariableEditor.spec.ts"
+test('custom variable query runner should return data when query is valid', async ({
+  variableEditPage,
+  page,
+  readProvisionedDataSource,
+  selectors,
+}) => {
+  const ds = await readProvisionedDataSource({ fileName: 'datasources.yaml' });
+  await variableEditPage.setVariableType('Query');
+  await variableEditPage.datasource.set(ds.name);
+  const codeEditorSelector = selectors.components.CodeEditor.container;
+  await variableEditPage.getByTestIdOrAriaLabel(codeEditorSelector).click();
+  await page.keyboard.insertText('select distinct(environment) from dataset');
+  const queryDataRequest = variableEditPage.waitForQueryDataRequest();
+  await variableEditPage.runQuery();
+  await queryDataRequest;
```

Comment:
What is happening under the hood here? Is there a reason why we are defining a variable and then awaiting that promise after the variabelEditPage.runQuery() is called?

Reply:
It's because it's the `runQuery` method that triggers the query data call. If we had awaited the request on line 64, we would never reach to line 65. On line 66, we halt execution until the query data request is completed, and then we can move on to asserting on what data is being displayed on the screen.

https://playwright.dev/docs/api/class-page#page-wait-for-request